apptest.go

```diff
@@ -357,6 +380,31 @@ func (a *AppSetup) waitForDeployedApp(ctx context.Context, appName string) error
 	return nil
 }
 
+// getCatalogURL returns the catalog URL for this app. If it is a Giant Swarm
+// catalog no URL needs to be provided.
+func getCatalogURL(app App) (string, error) {
+	if app.CatalogName == "" {
+		return "", microerror.Maskf(invalidConfigError, "catalog name must not be empty for app %#v", app)
+	}
+	if app.CatalogName != "" && app.CatalogURL != "" {
+		return app.CatalogURL, nil
+	}
+
+	for _, catalog := range giantSwarmCatalogs {
+		if app.CatalogName == catalog {
+			u, err := url.Parse(catalogBaseURL)
+			if err != nil {
+				return "", microerror.Mask(err)
+			}
+			u.Path = path.Join(u.Path, app.CatalogName)
+
+			return u.String(), nil
+		}
+	}
+
+	return "", microerror.Maskf(invalidConfigError, "catalog %#q not found and no URL provided", app.CatalogName)
```

Comment:
If `giantSwarmCatalogs` would be a map instead of slice, I think the code would be simpler

```suggestion
	_, exists := giantSwarmCatalogs[app.CatalogName]
	if !exists {
	    return "", microerror.Maskf(invalidConfigError, "catalog %#q not found and no URL provided", app.CatalogName)
	}

	u, err := url.Parse(catalogBaseURL)
	if err != nil {
		return "", microerror.Mask(err)
	}

	u.Path = path.Join(u.Path, app.CatalogName)
	return u.String(), nil
```

Reply:
Thanks Jose much better.

I also changed it to a string map. As some of the URLs are not as predictable as I thought.